lib/s3.py

```diff
@@ -111,14 +111,20 @@ def sign_curl(url: urllib.parse.ParseResult, method='GET', headers={}, checksum=
 
 def urlopen(url: urllib.parse.ParseResult, method='GET', headers={}, data=b'') -> IO:
     """Same as sign_request() but calls urlopen() on the result"""
-    headers = sign_request(url, method, headers, hashlib.sha256(data).hexdigest())
-    request = urllib.request.Request(url.geturl(), headers=headers, method=method, data=data)
-    try:
-        response = urllib.request.urlopen(request, context=host_ssl_context(url.netloc))
-    except urllib.error.HTTPError as exc:
-        logger.debug('%s %s %s → %s:', method, url.geturl(), headers, exc.status)
-        logger.debug('  %s', exc.read())
-        raise
+    tries = 3
+    while True:
+        headers = sign_request(url, method, headers, hashlib.sha256(data).hexdigest())
+        request = urllib.request.Request(url.geturl(), headers=headers, method=method, data=data)
+        try:
+            response = urllib.request.urlopen(request, context=host_ssl_context(url.netloc))
```

Comment:
This needs to be returned. Otherwise, it will indefinitely re-do the loop on success, eventually leading to 403.

Reply:
I thought I had that in there.  I wonder how it got dropped...

Reply:
Fixed.